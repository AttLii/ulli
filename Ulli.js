import React from "react"
import PropTypes from "prop-types"

export const Ulli = ({
  list,
  itemFunc,
  className,
  itemClassName,
  ItemElement,
  ListElement
}) => (
  list.length > 0 ? (
    <ListElement className={className}>
      {list.map((item, i) => (
        <ItemElement key={i} className={itemClassName}>
          {itemFunc(item, i)}
        </ItemElement>
      ))}
    </ListElement>
  ) : null
)

Ulli.defaultProps = {
  list: [],
  itemFunc: () => {},
  className: undefined,
  itemClassName: undefined,
  ListElement: "ul",
  ItemElement: "li"
}

Ulli.propTypes = {
  className: PropTypes.string,
  itemClassName: PropTypes.string,
  itemFunc: PropTypes.func,
  list: PropTypes.array,
  ListElement: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.func
  ]),
  ItemElement: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.func
  ]),
}