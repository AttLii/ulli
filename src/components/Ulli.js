import PropTypes from "prop-types"

export const Ulli = ({ className, itemClassName, itemFunc, list }) =>
  list.length > 0 ? (
    <ul className={className}>
      {list.map((item, i) => (
        <li key={i} className={itemClassName}>
          {itemFunc(item, i)}
        </li>
      ))}
    </ul>
  ) : null

Ulli.defaultProps = {
  className: undefined,
  itemFunc: () => {},
  itemClassName: undefined,
  list: []
}

Ulli.propTypes = {
  className: PropTypes.string,
  itemClassName: PropTypes.string,
  itemFunc: PropTypes.func,
  list: PropTypes.array
}