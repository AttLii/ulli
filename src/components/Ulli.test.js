import { render } from '@testing-library/react';
import renderer from "react-test-renderer"
import React from "react"

import { Ulli } from './Ulli';

test("doesn't crash", () => {
  render(<Ulli />);
});

test("doesn't render anything without list-prop", () => {
  const tree = renderer
    .create(<Ulli />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("doesn't render anything with empty list-prop", () => {
  const tree = renderer
    .create(<Ulli list={[]}/>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("renders only ul > li without itemFn-prop", () => {
  const tree = renderer
    .create(<Ulli list={["item1", "item2"]}/>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("renders itemFunc's result as each <li> elements children", () => {
  const itemFunc = () => "hi"
  
  const tree = renderer
    .create(
      <Ulli
        list={["item1", "item2"]}
        itemFunc={itemFunc}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("itemFunc has access to each item and current index", () => {
  const itemFunc = (item, i) => <div>{item} {i}</div>

  const tree = renderer
    .create(
      <Ulli
        list={["item1", "item2"]}
        itemFunc={itemFunc}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("component accepts className and itemClassName prop, that is applied to elements", () => {
  const itemFunc = (item, i) => <div>{item} {i}</div>

  const tree = renderer
    .create(
      <Ulli
        className="app__list"
        itemClassName="app__list-item"
        list={["item1", "item2"]}
        itemFunc={itemFunc}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});