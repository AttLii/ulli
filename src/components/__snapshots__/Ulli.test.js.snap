// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component accepts className and itemClassName prop, that is applied to elements 1`] = `
<ul
  className="app__list"
>
  <li
    className="app__list-item"
  >
    <div>
      item1
       
      0
    </div>
  </li>
  <li
    className="app__list-item"
  >
    <div>
      item2
       
      1
    </div>
  </li>
</ul>
`;

exports[`doesn't render anything with empty list-prop 1`] = `null`;

exports[`doesn't render anything without list-prop 1`] = `null`;

exports[`itemFunc has access to each item and current index 1`] = `
<ul>
  <li>
    <div>
      item1
       
      0
    </div>
  </li>
  <li>
    <div>
      item2
       
      1
    </div>
  </li>
</ul>
`;

exports[`renders itemFunc's result as each <li> elements children 1`] = `
<ul>
  <li>
    hi
  </li>
  <li>
    hi
  </li>
</ul>
`;

exports[`renders only ul > li without itemFn-prop 1`] = `
<ul>
  <li />
  <li />
</ul>
`;
